### ISSUE TITLE
ast Module

### ISSUE THREAD

**Issue by stewpend0us** [bjodah/sympy#6]
Does it make sense to make a class Module (or some other name) that would allow you to print collections of functions in addition to the #includes they may need? (this would hopefully obsolete util.codegen)

I'm not sure if it makes sense for other languages but for c something like:

m = Module( public_functions, private_functions) #the inputs are lists of FunctionDefinition
ccode(m.header)
ccode(m.source)

**Comment by bjodah** [comment]
Yes, I have started doing that, the details will differ depending on language:

- Fortran: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/futils.py
- Python: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/pyutils.py
- C: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/cutils.py

**Comment by bjodah** [comment]
And here are some tests:

- C: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/tests/test_applications.py#L27
- Fortran: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/tests/test_algorithms.py#L59
- Python: https://github.com/bjodah/sympy/blob/d313e132167a82070e3fb06e9cbf173937f1e9ab/sympy/codegen/tests/test_algorithms.py#L77

**Comment by stewpend0us** [comment]
Gotcha. Is the idea to eventually have ast nodes for #include, #ifndef? Or would there be custom functions in cutils.py to handle that?

**Comment by bjodah** [comment]
Gernerally it's a good idea to make things nodes if either:
- you could reasonably expect that someone would like to manipulate the ast using e.g. "replace"
- you're not sure if the node will always be at the top level (i.e. the printer needs to handle it when it comes across it at an arbitrary depth in a tree)

...or just for consistency's sake. For include I'm not sure point 1 & 2 holds, but maybe? :-)

**Comment by bjodah** [comment]
@stewpend0us, I'm just curious, so checking in to hear if you've found the time to use the AST nodes any more since March?

**Comment by stewpend0us** [comment]
Sorry...not much. I kinda stopped once I figured out how to get it to do what I needed. I wrote a couple helper functions to actually write out the .h and .c files with the correct #include statements and whatnot (not using util.codegen) and moved on. I don't think I have any real complaints about the AST nodes. (After a lot of reading and understanding) I was able to do what I needed to do and it worked great. Most of my issues were with other aspects of sympy like cse not playing nice sometimes and Matrices being a little...strange.
